Add support for context parallelism

bclyang opened the pull request:
Adds context parallelism with ring attention.

See WandB report for training runs to test correctness with a simple 410M config: https://wandb.ai/brandony/neox/reports/Test-context-parallelism-correctness--Vmlldzo5NTU4ODc1.
- Checked the following settings: no MP/CP, MP 4, CP 4, MP 2 and CP 2
- Confirmed that the loss exactly matches
- Memory and training speed seems reasonable

Based on this initial PR: https://github.com/EleutherAI/gpt-neox/pull/1266, with changes to get things working:
- All-reduce gradients across context-parallel nodes by piggy backing on DP
- Fix parallelism initialization
- Remove unnecessary code

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/gpt-neox?pullRequest=1299) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/gpt-neox?pullRequest=1299) before we can accept your contribution.<br/>**8** out of **9** committers have signed the CLA.<br/><br/>:white_check_mark: aflah02<br/>:white_check_mark: Quentin-Anthony<br/>:white_check_mark: bclyang<br/>:white_check_mark: emmanuel-ferdman<br/>:white_check_mark: DayOfThePenguin<br/>:white_check_mark: aurelion-source<br/>:white_check_mark: Robby955<br/>:white_check_mark: Mr-Neutr0n<br/>:x: Jacob Hatef<br/><hr/>**Jacob Hatef** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/gpt-neox?pullRequest=1299) it.</sub>

bclyang committed:
Add context parallelism support

bclyang committed:
Add context parallelism support

Quentin-Anthony commented on megatron/model/fused_layer_norm.py:
```diff
@@ -37,7 +37,7 @@ def __init__(
         normalized_shape,
         eps=1e-5,
         no_persist_layer_norm=True,
-        sequence_parallel=False,
```
@bclyang -- Shouldn't these remain `sequence_parallel` to match our previous support for megatron-style sequence parallelism (essentially just TP applied to layernorm and dropout)?

Quentin-Anthony committed:
Merge branch 'main' of https://github.com/EleutherAI/gpt-neox into add-context-parallel-support

aurelion-source committed:
Make returning moe_losses optional in SequentialWrapper (#1336)

* - Adds a flag to check whether to return moe losses in SequentialWrapper

* pre-commit

---------

Co-authored-by: Nawras Alnaasan <alnaasan.1@osu.edu>
Co-authored-by: Quentin Anthony <qganthony@yahoo.com>

aurelion-source committed:
Creating NonParametricLayernorm class (#1338)

* Added non_parametric_layernorm

* Added non_parametric_layernorm to args

* Create 7B.yml\

* Create 1B.yml

* Delete configs/OLMo directory

* Create new class for NonParametricLayernorm

* precommit

---------

Co-authored-by: Aflah <72096386+aflah02@users.noreply.github.com>
Co-authored-by: Quentin Anthony <qganthony@yahoo.com>

Quentin-Anthony committed:
remove deprecated nvlink arg (#1339)

Quentin-Anthony committed:
update deepspeed commit (#1343)

Quentin-Anthony committed:
fix te link (#1344)

aurelion-source committed:
Fix pretty_print in inspect_checkpoints.py (#1346)

Quentin-Anthony committed:
add apptainer file and docs (#1351)

aurelion-source committed:
Resolve Transformer Engine SwiGLU Issue (#1352)

* Fix SwiGLU Activation Handling in Transformer Engine

* precommit

---------

Co-authored-by: Quentin Anthony <qganthony@yahoo.com>

aurelion-source committed:
NeoX+TE --> HF Conversion  (#1355)

* - Removes redundant post_attention_layernorm initialization when using te_lnmlp
- Adds correct mapping for neox_te -> hf conversion

* NeoX+TE --> HF Conversion
- Removes redundant post_attention_layernorm initialization when using te_lnmlp
- Adds correct mapping for neox_te -> hf conversion

* Set default rotary-emb-base to 10000

aurelion-source committed:
Fix issues with Llama HF->NeoX conversion (#1345)

* - Fixes Hf llama -> neox conversion by simply concatenating q,k,b
  instead of splitting by heads first for GQA
- Fixes RMSNorm implementation by adding epsilon to the varience
  instead of adding it directly to RMS

* - Remove RMSNorm extra argument

* Fixes qkv projections for neox->hf conversion

* Fixes qkv projections for neox->hf conversion

* precommit

---------

Co-authored-by: Quentin Anthony <qganthony@yahoo.com>

DayOfThePenguin committed:
Dmoe integration (#1210)

* feat: ensure only the right architectures get built vs all of them for 11.x

* feat: clean up megablocks-based DMoE implementation

* feat: update args, configs, and requirements

* Update NeoXArgs docs automatically

* Update NeoXArgs docs automatically

* feat: Update readme and example config

* Update NeoXArgs docs automatically

* Update NeoXArgs docs automatically

* - Updated sinkhorn initialization and add max_iter argument.
- Removed mp assertion for moe
- Removed mlp_type checks in moe code
- Added Bf16 conversion to dmoe_gather

* - Updated dmoe config
    - Fixed merge issues
    - Fixed moe sequence parallelism bug
    - Added assertions for zero stages 2 and 3 with moe
    - Updated moe requirements

* - Updated transformer_engine requirments
- Import megablocks only if num_experts > 1

* - Disable torch.compile explicitly when using TE

---------

Co-authored-by: github-actions <github-actions@github.com>
Co-authored-by: Quentin Anthony <qganthony@yahoo.com>
Co-authored-by: aurelion-source <nonnawras@gmail.com>

emmanuel-ferdman committed:
Fix neoxargs consume without config dir test (#1357)

Signed-off-by: Emmanuel Ferdman <emmanuelferdman@gmail.com>

Quentin-Anthony committed:
update and unify copyrights (#1360)

* update copyrights

* update remaining copyrights

aflah02 committed:
Add B200 and TE Images with Usage Notes (#1385)

* Add B200 and TE Images with Notes

* formatting

* cleanup

* formatting on dockerfile

---------

Co-authored-by: Quentin Anthony <qganthony@yahoo.com>

aflah02 committed:
Fix TE Crash with RMSNorm (#1384)

aflah02 committed:
Add route to modify megatron fused kernel build path (#1383)

* Add arg to configure build path

* cleanup and formatting

---------

Co-authored-by: Quentin Anthony <qganthony@yahoo.com>

aflah02 committed:
Add assert to catch case when neither epochs or iters is set (#1368)

Mr-Neutr0n committed:
fix: correct inverted early-stop condition in Sinkhorn iteration (#1370)

Robby955 committed:
Fix warmup LR being zero at step 0 (#1379)

At step 0, num_iters is 0 so the warmup formula
`start_lr * num_iters / warmup_iter` yields LR=0.
This means the very first training step is a no-op:
the step-1 checkpoint is identical to the step-0 checkpoint.

Fix by using max(1, num_iters) in the warmup formula so
step 0 gets the same small nonzero LR that step 1 would
have received (start_lr / warmup_iter).

Fixes #1373

Mr-Neutr0n committed:
fix: gMLP uses full bias instead of truncated bias (#1371)

* fix: correct undefined self.args in TopKTokenChoiceRouter

* fix: use correctly sliced bias in gMLP projection

Commit:
merge
